blockdag/process.go

```diff
@@ -219,69 +118,339 @@ func (dag *BlockDAG) processBlockNoLock(block *util.Block, flags BehaviorFlags)
 		if err != nil {
 			return false, false, err
 		}
-		return false, true, err
+		return false, true, nil
 	}
 
 	// Handle orphan blocks.
 	if len(missingParents) > 0 {
-		// Some orphans during netsync are a normal part of the process, since the anticone
-		// of the chain-split is never explicitly requested.
-		// Therefore, if we are during netsync - don't report orphans to default logs.
-		//
-		// The number K*2 was chosen since in peace times anticone is limited to K blocks,
-		// while some red block can make it a bit bigger, but much more than that indicates
-		// there might be some problem with the netsync process.
-		if flags&BFIsSync == BFIsSync && dagconfig.KType(len(dag.orphans)) < dag.Params.K*2 {
-			log.Debugf("Adding orphan block %s. This is normal part of netsync process", blockHash)
-		} else {
-			log.Infof("Adding orphan block %s", blockHash)
-		}
 		dag.addOrphanBlock(block)
-
 		return true, false, nil
 	}
+	return false, false, nil
+}
 
-	// The block has passed all context independent checks and appears sane
-	// enough to potentially accept it into the block DAG.
-	err = dag.maybeAcceptBlock(block, flags)
+func (dag *BlockDAG) processOrphansAndDelayedBlocks(blockHash *daghash.Hash, flags BehaviorFlags) error {
+	err := dag.processOrphans(blockHash, flags)
 	if err != nil {
-		return false, false, err
+		return err
+	}
+
+	if !isBehaviorFlagRaised(flags, BFAfterDelay) {
+		err = dag.processDelayedBlocks()
+		if err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	// Accept any orphan blocks that depend on this block (they are
-	// no longer orphans) and repeat for those accepted blocks until
-	// there are no more.
-	err = dag.processOrphans(blockHash, flags)
+// maybeAcceptBlock potentially accepts a block into the block DAG. It
+// performs several validation checks which depend on its position within
+// the block DAG before adding it. The block is expected to have already
+// gone through ProcessBlock before calling this function with it.
+//
+// The flags are also passed to checkBlockContext and connectBlock. See
+// their documentation for how the flags modify their behavior.
+//
+// This function MUST be called with the dagLock held (for writes).
+func (dag *BlockDAG) maybeAcceptBlock(block *util.Block, flags BehaviorFlags) error {
+	err := dag.checkBlockContext(block, flags)
```

Comment:
I'm not 100% convinced that checkBlockContext should be in maybeAcceptBlock, and not in processBlock right before it.

TBH the distinction between processBlock and maybeAcceptBlock is not very clear.
One possible distinction is that maybeAcceptBlock is handling all the blockNode stuff. Therefore checkBlockContext should be outside.

LMK what you think.

Reply:
The only reason maybeAcceptBlock still exists is because it's used inside of processOrphans.
Ideally processOrphans would use processBlock (as processDelayedBlocks does, for example), but that would require a change that's outside the scope of this ticket.